File: leetcode/341.扁平化嵌套列表迭代器/341.扁平化嵌套列表迭代器.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
# class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """
"""
__init__ : 用栈进行储存
hasNext : 完成列表节点的展开
    - 如果是integer直接返回true
    - 如果是列表，就写入栈
    - 为空时返回False
next : 返回栈顶数据

"""


class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.stack = []
        # 从尾到头遍历
        for i in range(len(nestedList) - 1, -1, -1):
            self.stack.append(nestedList[i])

    def next(self) -> int:
        cur = self.stack.pop()
        return cur.getInteger()

    def hasNext(self) -> bool:
        while self.stack:
            cur = self.stack[-1]
            if cur.isInteger():
                return True
            self.stack.pop()
            for i in range(len(cur.getList()) - 1, -1, -1):
                self.stack.append(cur.getList[i])
        return False

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: leetcode/341.扁平化嵌套列表迭代器/341.扁平化嵌套列表迭代器.go
package main

//type NestedIterator struct {
//	Stack []*NestedInteger
//}
//
//func Constructor(nestedList []*NestedInteger) *NestedIterator {
//	stack := []*NestedInteger{}
//	for i := len(nestedList) - 1; i >= 0; i-- {
//		stack = append(stack, nestedList[i])
//	}
//	return &NestedIterator{Stack: stack}
//}
//
//func (this *NestedIterator) Next() int {
//	cur := this.Stack[len(this.Stack)-1]
//	this.Stack = this.Stack[:len(this.Stack)-1]
//	return cur.GetInteger()
//}
//
//func (this *NestedIterator) HasNext() bool {
//	for len(this.Stack) > 0 {
//		cur := this.Stack[len(this.Stack)-1]
//		if cur.IsInteger() {
//			return true
//		}
//		this.Stack = this.Stack[:len(this.Stack)-1]
//		list := cur.GetList()
//		for i := len(list) - 1; i >= 0; i-- {
//			this.Stack = append(this.Stack, list[i])
//
//		}
//	}
//	return false
//}
